tg/tg_module.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2024 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package tg
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"slices"
+	"sort"
+
+	"github.com/gruntwork-io/go-commons/env"
+	"github.com/gruntwork-io/terragrunt/config"
+	"github.com/gruntwork-io/terragrunt/configstack"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/rs/zerolog/log"
+	"github.com/terramate-io/terramate/errors"
+	"github.com/terramate-io/terramate/project"
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+// ErrParsing indicates there is an error parsing a Terragrunt file.
+const ErrParsing errors.Kind = "parsing Terragrunt file"
+
+type (
+	// Module is a Terragrunt module.
+	Module struct {
+		Path       project.Path  `json:"path"`
+		ConfigFile project.Path  `json:"config"`
+		DependsOn  project.Paths `json:"depends_on"`
+	}
+
+	// Modules is a list of Module.
+	Modules []*Module
+)
+
+// ScanModules scans dir looking for Terragrunt modules. It returns a list of
+// modules with its "trigger paths" computed. The TriggerPaths are paths that,
+// when changed, must mark the module as changed.
+func ScanModules(rootdir string, dir project.Path) (Modules, error) {
+	absDir := project.AbsPath(rootdir, dir.String())
+	opts := newTerragruntOptions(absDir)
+
+	tgConfigFiles, err := config.FindConfigFilesInPath(absDir, opts)
+	if err != nil {
+		return nil, errors.E(err, "scanning Terragrunt modules")
+	}
+
+	logger := log.With().
+		Str("action", "tg.ScanModules").
+		Logger()
+
+	var modules Modules
+
+	sort.Strings(tgConfigFiles)
+	slices.Compact(tgConfigFiles)
+
+	fileErrs := map[string]*errors.List{}
+	fileProcessed := map[string]struct{}{}
+	for _, cfgfile := range tgConfigFiles {
+		fileErrs[cfgfile] = errors.L()
+		logger := logger.With().Str("cfg-file", cfgfile).Logger()
+
+		logger.Trace().Msg("found configuration")
+
+		cfgOpts := opts.Clone(cfgfile)
+
+		pctx := config.NewParsingContext(context.Background(), cfgOpts).WithDecodeList(
+			// needed for tracking:
+			//   - terraform.extra_arguments
+			//   - terraform.required_vars_file
+			//   - terraform.optional_var_files
+			//   - etc
+			config.TerraformBlock,
+
+			// Needed for detecting modules.
+			config.TerraformSource,
+
+			// Need for parsing out the dependencies
+			config.DependencyBlock,
+		)
+
+		mod := &Module{
+			Path:       project.PrjAbsPath(rootdir, cfgOpts.WorkingDir),
+			ConfigFile: project.PrjAbsPath(rootdir, cfgfile),
+		}
+
+		pctx.PredefinedFunctions = make(map[string]function.Function)
+		pctx.PredefinedFunctions[config.FuncNameFindInParentFolders] = findInParentFoldersFunc(pctx, rootdir, mod)
+		pctx.PredefinedFunctions[config.FuncNameReadTerragruntConfig] = readTerragruntConfigFunc(pctx, rootdir, mod)
+		pctx.PredefinedFunctions[config.FuncNameReadTfvarsFile] = wrapStringSliceToStringAsFuncImpl(pctx, rootdir, mod, readTFVarsFile)
+
+		tgConfig, err := config.PartialParseConfigFile(
```

Comment:
Maybe we could add some comments here? If I understand correctly, we replace some of the predefined functions, then we call terragrunt a terragrunt function, which will call our hooks, where we collect the data we are interested it, then we read that data out after the terragrunt function is done.

Where the hooks are added is clear, but I'd add comments to indicate where we call the terragrunt functions and when we have all the results we need.

Reply:
Will add a comment. The user's code is that calls the Terragrunt functions, we just intercept these calls and collect their referenced paths (if any).